### PR TITLE
Add support for overriding IdP Entity ID per SAML SP

### PR DESCRIPTION
**Pull request by DinikaSen**
### Proposed changes in this pull request

Add 2 SAML Service Provider configuration attributes to enable IdP Entity ID overriding support for service providers and to enable using an issuer qualifier to create multiple SAML SP configurations with the similar issuer. Modified unit tests in SAMLSSOServiceProviderDAOTest.

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java [hunk 1]
```diff
@@ -251,12 +271,24 @@ public boolean addServiceProvider(SAMLSSOServiceProviderDO serviceProviderDO) th
             }
             registry.put(path, resource);
             if (log.isDebugEnabled()) {
-                log.debug("Service Provider " + serviceProviderDO.getIssuer() + " is added successfully.");
+                if (StringUtils.isNotBlank(serviceProviderDO.getIssuerQualifier())) {
+                    log.debug("Service Provider " + serviceProviderDO.getIssuer() + " with issuer "
```
Comment (omindu) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java]: Change Service Provider to SAML2 Service Provider

Comment (DinikaSen) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java]: Changed

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAOTest.java [hunk 2]
```diff
@@ -172,6 +173,15 @@ private void setUpResources() throws Exception {
...
         dummyAdvProperties.put(IdentityRegistryResources.PROP_SAML_ENABLE_ECP, Collections
                 .singletonList("true"));
+        dummyAdvProperties.put(IdentityRegistryResources.PROP_SAML_SSO_IDP_ENTITY_ID_ALIAS, Collections
+                .singletonList("dummyIdPEntityAlias"));
+
+        dummyPropertiesWithAnIssuerQualifier = new HashMap<>();
+        dummyPropertiesWithAnIssuerQualifier.putAll(dummyBasicProperties);
+        dummyPropertiesWithAnIssuerQualifier.put(IdentityRegistryResources.PROP_SAML_SSO_ISSUER, Collections.singletonList
+                ("DummyIssuer"));
+        dummyPropertiesWithAnIssuerQualifier.put(IdentityRegistryResources.PROP_SAML_SSO_ISSUER_QUALIFIER, Collections.singletonList
```
[the remark this answers is not in the document]
Comment (DinikaSen) [reply components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAOTest.java]: reformatted

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java [hunk 3]
```diff
@@ -575,6 +641,11 @@ public SAMLSSOServiceProviderDO uploadServiceProvider(SAMLSSOServiceProviderDO s
             throw new IdentityException("Issuer cannot be found in the provided arguments.");
         }
 
+        if (StringUtils.isNotBlank(serviceProviderDO.getIssuerQualifier())) {
```
Comment (denuwanthi) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java]: This code block and code block at https://github.com/wso2/carbon-identity-framework/pull/1999/files#diff-a097c8a250ffb40fa9541d765fac50deR244 seems to be same.
If so, we can move it to a separate method and reuse.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java [hunk 4]
```diff
@@ -174,6 +176,26 @@ public void setIssuer(String issuer) {
         }
     }
 
+    /**
+     * Get qualifier of the issuer.
+     *
+     * @return Issuer Qualifier
+     */
+    public String getIssuerQualifier() {
+        return issuerQualifier;
```
Comment (ashensw) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java:185]: Add a new line after the method signature.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java [hunk 5]
```diff
@@ -174,6 +176,26 @@ public void setIssuer(String issuer) {
...
+    }
+
+    /**
+     * Set issuer qualifier.
+     *
+     * @param issuerQualifier
+     */
+    public void setIssuerQualifier(String issuerQualifier) {
+        if (StringUtils.isNotBlank(issuerQualifier)) {
```
Comment (ashensw) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java:194]: Add a new line after the method signature.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java [hunk 6]
```diff
@@ -590,4 +612,21 @@ public boolean isSamlECP() {
         return samlECP;
     }
 
+    /**
+     * Get IdP Entity ID alias.
+     *
+     * @return IdP Entity ID Alias
+     */
+    public String getIdpEntityIDAlias() {
+        return idpEntityIDAlias;
```
Comment (ashensw) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java:621]: Add a new line after the method signature.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java [hunk 7]
```diff
@@ -590,4 +612,21 @@ public boolean isSamlECP() {
...
+    }
+
+    /**
+     * Set IdP Entity ID alias.
+     *
+     * @param idpEntityIDAlias
+     */
+    public void setIdpEntityIDAlias(String idpEntityIDAlias) {
+        this.idpEntityIDAlias = idpEntityIDAlias;
```
Comment (ashensw) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java:630]: Add a new line after the method signature.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java [hunk 8]
```diff
@@ -232,15 +240,27 @@ public boolean addServiceProvider(SAMLSSOServiceProviderDO serviceProviderDO) th
             throw new IdentityException("Issuer cannot be found in the provided arguments.");
         }
 
+        // If an issuer qualifier value is specified, it is appended to the end of the issuer value
```
Comment (sachiniWettasinghe) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java:243]: `// If an issuer qualifier value is specified, it is appended to the end of the issuer value.`

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java (hunk 4 above)
Comment (DinikaSen) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java:185]: Addressed in https://github.com/wso2/carbon-identity-framework/pull/3100

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java (hunk 5 above)
Comment (DinikaSen) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java:194]: Addressed in https://github.com/wso2/carbon-identity-framework/pull/3100

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java (hunk 6 above)
Comment (DinikaSen) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java:621]: Addressed in https://github.com/wso2/carbon-identity-framework/pull/3100

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java (hunk 7 above)
Comment (DinikaSen) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/SAMLSSOServiceProviderDO.java:630]: Addressed in https://github.com/wso2/carbon-identity-framework/pull/3100

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java (hunk 8 above)
Comment (DinikaSen) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java:243]: Addressed in https://github.com/wso2/carbon-identity-framework/pull/3100

**Pull request merged**